### PR TITLE
Update [ConditionalFact] certificate detectors to install and check

### PR DESCRIPTION
**Pull request by roncain**
Changes the certificate [ConditionalFact] detectors to attempt to
install the certificates but then also to verify that they exist
in the store.  This permits scenarios where the install can fail
but a certificate is already in the store.

This change also renames and refactors BridgeClientCertificateManager
to remove all unused Bridge code and to simplify for only the certificate
operations currently required.

### CODE REVIEW

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs [hunk 1]
```diff
@@ -1,350 +0,0 @@
-﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: Reviewers: BridgeClientCertificateManager was renamed to CertificateManager and then modified.  I'll note what is new in CertificateManager and what was not retained from BridgeClientCertificateManager.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs [hunk 2]
```diff
@@ -1,350 +0,0 @@
...
-        private const string ClientCertificateSubject = "WCF Client Certificate";
-        private const string ClientCertificatePassword = "test"; // this needs to be kept in sync with the Bridge configuration 
-
-        public static string LocalCertThumbprint { get; set; }
-
-        // Dictionary of certificates installed by CertificateManager
-        // Keyed by the Thumbprint of the certificate
-        private static Dictionary<string, CertificateCacheEntry> s_myCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
-        private static Dictionary<string, CertificateCacheEntry> s_rootCertificates = new Dictionary<string, CertificateCacheEntry>(StringComparer.OrdinalIgnoreCase);
-
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: Reviewers: all these want away in the refactor/move

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs [hunk 3]
```diff
@@ -1,350 +0,0 @@
...
-                }
-                return s_PlatformSpecificRootStoreLocation;
-            }
-        }
-
-        // Installs the root certificate from the bridge 
-        // Needed for all tests so we trust the incoming certificate coming from the service/bridge
-        // We do our best to detect if we're running on the same box as the bridge; if so, we don't try to install the cert
-        // as that operation requires admin privileges
-        public static void InstallRootCertificateFromBridge()
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs]: Reviewers: all methods containing the name Bridge went away.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        // See dotnet/corefx#3690
+        //
+        // For Windows: 
+        // We don't want to use CurrentUser, as writing to StoreName.Root : StoreLocation.CurrentUser will result in a 
+        // modal dialog box that isn't dismissable if the root cert hasn't already been installed previously. 
+        // If the cert has already been installed, writing to StoreName.Root : StoreLocation.CurrentUser results in a no-op
+        // 
+        // In other words, on Windows, we can bypass the modal dialog box, but only if we install to StoreName.Root : StoreLocation.LocalMachine
+        // To do this though means that we must run certificate-based tests elevated
+        private static StoreLocation PlatformSpecificRootStoreLocation
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs]: Reviewers: this came verbatim from BridgeClientCertificateManager

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                    s_PlatformSpecificStoreLocationIsSet = true;
+                }
+                return s_PlatformSpecificRootStoreLocation;
+            }
+        }
+
+        // Adds the given certificate to the given store unless it is
+        // already present.  Returns the  certificate either already in
+        // the store or the one requested.
+        public static X509Certificate2 AddToStoreIfNeeded(StoreName storeName,
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs]: Reviewers: almost verbatim from BridgeClientCertificateManager.  I removed the separate cache we were maintaining and made it return the cert actually in the store if one was already present.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+        }
+
+        // Retrieves a root certificate matching the given thumbprint from the root store
+        public static X509Certificate2 RootCertificateFromThumprint(string thumbprint)
+        {
+            return CertificateFromThumbprint(StoreName.Root, PlatformSpecificRootStoreLocation, thumbprint);
+        }
+
+        // Retrieves a root certificate matching the given issuer name and subject name from the root store
+        public static X509Certificate2 RootCertificateFromName(string issuerName, string subjectName)
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs]: The "FromName" methods are new and exist only to support the final check used by the detectors if the install fails.

File: src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                    return null;
+                }
+            }
+        }
+
+
+        // Install the certificate into the Root store and returns its thumbprint.
+        // It will not install the certificate if it is already present in the store.
+        // It returns the thumbprint of the certificate, regardless whether it was added or found.
+        public static X509Certificate2 InstallCertificateToRootStore(X509Certificate2 certificate)
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs]: Reviewers: mostly the same but removed all local caching the old code was doing.

File: src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Https/HttpsTests.cs [hunk 8]
```diff
@@ -184,7 +184,7 @@ public static void ServerCertificateValidation_EchoString()
             CustomBinding binding = new CustomBinding(new TextMessageEncodingBindingElement(MessageVersion.Soap11, Encoding.UTF8), new HttpsTransportBindingElement());
 
             endpointAddress = new EndpointAddress(new Uri(Endpoints.Https_DefaultBinding_Address));
-            clientCertThumb = ServiceUtilHelper.LocalCertThumbprint;
+            clientCertThumb = ServiceUtilHelper.ClientCertificate.Thumbprint;
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Security/TransportSecurity/Https/HttpsTests.cs]: Reviewers: ServiceUtilHelper.ClientCertificate is guaranteed to be non-null if Client_Certificate_Installed is declared in [ConditionalFact].

**Comment by roncain** [comment]
@dotnet-bot test outerloop Windows_NT

**Comment by roncain** [comment]
@dotnet-bot test outerloop selfhost Windows_NT

**Comment by StephenBonikowsky** [comment]
LGTM, don't block merge waiting for feedback, not a product related change.

**Comment by roncain** [comment]
Rebasing and manually merging due to merge conflicts with CodeFormatter pass.  Will re-run OuterLoops afterwards

**Comment by roncain** [comment]
@dotnet-bot test outerloop selfhost Windows_NT

**Comment by roncain** [comment]
@dotnet-bot test outerloop Windows_NT

**Pull request merged**